batch2/batch/semaphore.py

```diff
@@ -23,24 +24,33 @@ def __init__(self, sem, weight):
         await self.sem.acquire(self.weight)
 
     async def __aexit__(self, exc_type, exc, tb):
-        await self.sem.release(self.weight)
+        self.sem.release(self.weight)
 
 
-class WeightedSemaphore:
+class FIFOWeightedSemaphore:
     def __init__(self, value=1):
         self.value = value
-        self.cond = asyncio.Condition()
+        self.queue = collections.deque()
 
     async def acquire(self, weight):
-        while self.value < weight:
-            async with self.cond:
-                await self.cond.wait()
-        self.value -= weight
+        if self.value >= weight:
+            self.value -= weight
+            return
```

Comment:
Couldn't this mean you're jumping the queue if someone else is waiting?  E.g. there are 6 cores free, someone is waiting for 8, and you only need one.

Reply:
good catch!

Reply:
This code is still there?  Did you maybe forget to push?